src/execution_plan/ops/op_update.c

```diff
@@ -172,33 +194,54 @@ OpBase* NewUpdateOp(GraphContext *gc, AST *ast, ResultSet *result_set) {
     op_update->op.consume = OpUpdateConsume;
     op_update->op.reset = OpUpdateReset;
     op_update->op.free = OpUpdateFree;
+    op_update->op.init = OpUpdateInit;
 
     return (OpBase*)op_update;
 }
 
+OpResult OpUpdateInit(OpBase *opBase) {
+    OpUpdate* op = (OpUpdate*)opBase;
+    if(_ShouldCacheRecord(op)) op->records = array_new(Record, 64);    
+    return OP_OK;
+}
+
 Record OpUpdateConsume(OpBase *opBase) {
     OpUpdate *op = (OpUpdate*)opBase;
     OpBase *child = op->op.children[0];
-    Record r = child->consume(child);
-    // No data, return.
-    if(!r) return NULL;
-
-    /* Evaluate each update expression and store result 
-     * for later execution. */
-    EntityUpdateEvalCtx *update_expression = op->update_expressions;
-    for(uint i = 0; i < op->update_expressions_count; i++, update_expression++) {
-        SIValue new_value = AR_EXP_Evaluate(update_expression->exp, r);
-
-        // Make sure we're updating either a node or an edge.
-        RecordEntryType t = Record_GetType(r, update_expression->entityRecIdx);
-        assert(t == REC_TYPE_NODE || t == REC_TYPE_EDGE);
-        GraphEntityType type = (t == REC_TYPE_NODE) ? GETYPE_NODE : GETYPE_EDGE;
-
-        GraphEntity *entity = Record_GetGraphEntity(r, update_expression->entityRecIdx);
-        _QueueUpdate(op, entity, type, update_expression->attribute, new_value);
+    Record r;
+
+    // Updates already performed.
+    if(op->updates_commited) return _handoff(op);
+
+    while((r = child->consume(child))) {
+        /* Evaluate each update expression and store result 
+         * for later execution. */
+        EntityUpdateEvalCtx *update_expression = op->update_expressions;
+        for(uint i = 0; i < op->update_expressions_count; i++, update_expression++) {
+            SIValue new_value = AR_EXP_Evaluate(update_expression->exp, r);
+
+            // Make sure we're updating either a node or an edge.
+            RecordEntryType t = Record_GetType(r, update_expression->entityRecIdx);
+            assert(t == REC_TYPE_NODE || t == REC_TYPE_EDGE);
+            GraphEntityType type = (t == REC_TYPE_NODE) ? GETYPE_NODE : GETYPE_EDGE;
+
+            GraphEntity *entity = Record_GetGraphEntity(r, update_expression->entityRecIdx);
+            _QueueUpdate(op, entity, type, update_expression->attribute, new_value);
+        }
+
+        if(_ShouldCacheRecord(op)) op->records = array_append(op->records, r);
+        // Record not going to be used, discard.
+        else Record_Free(r);
```

Comment:
As a style preference, I don't like having naked `if` and `else` statements not directly contiguous - can we use curly braces or put the comment after the expression on 234?